crates/router/src/core/payments/helpers.rs

```diff
@@ -2488,3 +2488,87 @@ mod test {
         );
     }
 }
+
+pub async fn get_additional_payment_data(
+    pm_data: &api_models::payments::PaymentMethodData,
+    db: &dyn StorageInterface,
+) -> api_models::payments::AdditionalPaymentData {
+    match pm_data {
+        api_models::payments::PaymentMethodData::Card(card_data) => {
+            if card_data.card_issuer.is_some()
+                && card_data.card_network.is_some()
+                && card_data.card_type.is_some()
+                && card_data.card_issuing_country.is_some()
+                && card_data.bank_code.is_some()
+            {
+                api_models::payments::AdditionalPaymentData::Card {
+                    card_issuer: card_data.card_issuer.to_owned(),
+                    card_network: card_data.card_network.clone(),
+                    card_type: card_data.card_type.to_owned(),
+                    card_issuing_country: card_data.card_issuing_country.to_owned(),
+                    bank_code: card_data.bank_code.to_owned(),
+                }
+            } else {
+                let card_number = card_data.clone().card_number;
+                let card_info = db
+                    .get_card_info(&card_number.get_card_isin())
+                    .await
+                    .map_err(|error| services::logger::error!(%error, "Failed to retrieve card information"))
+                    .transpose()
+                    .and_then(|card_info| (card_info).ok().map(|card_info| api_models::payments::AdditionalPaymentData::Card {
+                               card_issuer: card_info.card_issuer.as_ref().map(|issuer| issuer.to_string()),
+                                card_network: card_info.card_network.clone().map(|network| network.foreign_into()),
+                                bank_code: card_info.bank_code.as_ref().map(|bank_code| bank_code.to_string()),
+                                 card_type: card_info.card_type.as_ref().map(|card_type| card_type.to_string()),
+                               card_issuing_country: card_info.card_issuing_country.as_ref().map(|card_issuing_country| card_issuing_country.to_string()),
+                        }));
```

Comment:
I think using flatten would have been cleaner.
```suggestion
                    .map_err(|error| services::logger::warn!(card_info_error=?error))
                    .ok()
                    .flatten()
                    .map(
                        |card_info| api_models::payments::AdditionalPaymentData::Card {
                            card_issuer: card_info.card_issuer,
                            card_network: card_info
                                .card_network
                                .clone()
                                .map(|network| network.foreign_into()),
                            bank_code: card_info.bank_code,
                            card_type: card_info.card_type,
                            card_issuing_country: card_info.card_issuing_country,
                        },
                    );
```

Reply:
done!